### PR TITLE
[GOBBLIN-1613] Add metadata writers field to GMCE schema

### PR DESCRIPTION
**Pull request by homatthew**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [X] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-1613] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1613

### Description
- [X] Here are some details about my PR, including screenshots (if applicable):

Add optional field to GMCE schema for specifying metadata writers that should write. I've added corresponding code to filter base on this metadata writer field the GobblinMCEWriter

### Tests
- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Unit tests added and tested on holdem using iceberg metadata repair job

### Commits
- [X] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/GobblinMCEWriter.java [hunk 1]
```diff
@@ -294,12 +298,14 @@ public void writeEnvelope(RecordEnvelope<GenericRecord> recordEnvelope) throws I
   }
 
   // Add fault tolerant ability and make sure we can emit GTE as desired
-  private void write(RecordEnvelope recordEnvelope, ConcurrentHashMap newSpecsMap, ConcurrentHashMap oldSpecsMap, HiveSpec spec) throws IOException {
+  @VisibleForTesting
+  void write(RecordEnvelope<GenericRecord> recordEnvelope, ConcurrentHashMap newSpecsMap, ConcurrentHashMap oldSpecsMap, HiveSpec spec) throws IOException {
```
Comment (homatthew) [inline gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/GobblinMCEWriter.java:302]: Entry point for calling writers

File: gobblin-iceberg/src/test/java/org/apache/gobblin/iceberg/writer/GobblinMCEWriterTest.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+  public void testWriteNoOpIfNotSpecified() throws IOException {
+    gmceBuilder.setAllowedMetadataWriters(Arrays.asList("Some other metadata writer"));
+    gobblinMCEWriter.write(
+        new RecordEnvelope<>(gmceBuilder.build(), watermark),
+        new ConcurrentHashMap(), new ConcurrentHashMap(), mockHiveSpec);
+
+    Mockito.verify(mockWriter, never()).writeEnvelope(any(), any(), any(), any());
+  }
+
+  @Test(dataProvider = "AllowMockMetadataWriter")
```
Comment (homatthew) [inline gobblin-iceberg/src/test/java/org/apache/gobblin/iceberg/writer/GobblinMCEWriterTest.java:141]: We expect calls to the mock writer if the class name is specified in the GMCE or there are no metadata writers specified.

File: gobblin-iceberg/build.gradle [hunk 3]
```diff
@@ -56,6 +56,7 @@ dependencies {
     testCompile project(path: ':gobblin-modules:gobblin-kafka-common', configuration: 'tests')
     testCompile externalDependency.testng
     testCompile externalDependency.mockito
+    testCompile externalDependency.powermock
```
Comment (homatthew) [inline gobblin-iceberg/build.gradle:59]: Mocking static method. Need to use this to mock the GobblinConstructorUtils and the FileSystem static calls

File: gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/GobblinMCEWriter.java [hunk 4]
```diff
@@ -294,12 +298,14 @@ public void writeEnvelope(RecordEnvelope<GenericRecord> recordEnvelope) throws I
   }
 
   // Add fault tolerant ability and make sure we can emit GTE as desired
-  private void write(RecordEnvelope recordEnvelope, ConcurrentHashMap newSpecsMap, ConcurrentHashMap oldSpecsMap, HiveSpec spec) throws IOException {
+  @VisibleForTesting
```
Comment (homatthew) [inline gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/GobblinMCEWriter.java:301]: The write envelope method that calls this method has complicated logic related to hive that is unrelated to this change. So it makes sense to test this internal method instead of the public `writeEnvelope` method.

File: gobblin-iceberg/build.gradle (hunk 3 above)
Comment (ZihanLi58) [reply gobblin-iceberg/build.gradle:59]: Can you add these comments to code as well? so it's easily for us to track and understand in the future

File: gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/GobblinMCEWriter.java [hunk 5]
```diff
@@ -294,12 +298,14 @@ public void writeEnvelope(RecordEnvelope<GenericRecord> recordEnvelope) throws I
...
+  void write(RecordEnvelope<GenericRecord> recordEnvelope, ConcurrentHashMap newSpecsMap, ConcurrentHashMap oldSpecsMap, HiveSpec spec) throws IOException {
     boolean meetException = false;
     String dbName = spec.getTable().getDbName();
     String tableName = spec.getTable().getTableName();
     String tableString = Joiner.on(TABLE_NAME_DELIMITER).join(dbName, tableName);
-    for (MetadataWriter writer : metadataWriters) {
+    List<MetadataWriter> allowedWriters = getAllowedMetadataWriters(toGMCE(recordEnvelope), metadataWriters);
```
Comment (ZihanLi58) [inline gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/GobblinMCEWriter.java:307]: Can we avoid call toGMCE again in this method? As deepCopy is actually expansive..

File: gobblin-iceberg/src/test/java/org/apache/gobblin/iceberg/writer/GobblinMCEWriterTest.java [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.iceberg.writer;
+
```
Comment (ZihanLi58) [inline gobblin-iceberg/src/test/java/org/apache/gobblin/iceberg/writer/GobblinMCEWriterTest.java:19]: Thanks for adding this test job!

File: gobblin-iceberg/src/test/java/org/apache/gobblin/iceberg/writer/GobblinMCEWriterTest.java [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    gobblinMCEWriter.close();
+  }
+
+  @BeforeMethod
+  public void setUp() throws Exception {
+    MockitoAnnotations.initMocks(this);
+    gmceBuilder = GobblinMetadataChangeEvent.newBuilder()
+        .setDatasetIdentifier(DatasetIdentifier.newBuilder()
+            .setDataPlatformUrn("urn:li:dataPlatform:hdfs")
+            .setNativeName("/data/tracking/testTable")
```
Comment (ZihanLi58) [inline gobblin-iceberg/src/test/java/org/apache/gobblin/iceberg/writer/GobblinMCEWriterTest.java:96]: Change a test path and test platform as this seems specific to LinkedIn?

**Review by ZihanLi58 — CHANGES_REQUESTED** [review]

File: gobblin-iceberg/src/test/resources/publisherTest/data.json [hunk 8]
```diff
@@ -1,8 +1,10 @@
...
-  "id": 1,
-  "name": "Alyssa"
-}
-{
-  "id": 2,
-  "name": "Bob"
-}
\ No newline at end of file
+[
+  {
```
Comment (homatthew) [inline gobblin-iceberg/src/test/resources/publisherTest/data.json:2]: Cleaning up "invalid" json syntax. I think the reason this worked before is because of Jackson magic

File: gobblin-iceberg/src/test/java/org/apache/gobblin/iceberg/writer/IcebergMetadataWriterTest.java [hunk 9]
```diff
@@ -346,7 +349,10 @@ public void testFaultTolerant() throws Exception {
     Mockito.doThrow(new IOException("Test failure")).when(mockWriter).writeEnvelope(Mockito.any(), Mockito.any(), Mockito.any(), Mockito.any());
     gobblinMCEWriter.metadataWriters.add(0, mockWriter);
 
-    GenericRecord genericGmce = GenericData.get().deepCopy(gmce.getSchema(), gmce);
+    GobblinMetadataChangeEvent gmceWithMockWriter = SpecificData.get().deepCopy(gmce.getSchema(), gmce);
+    gmceWithMockWriter.setAllowedMetadataWriters(Arrays.asList(IcebergMetadataWriter.class.getName(), mockWriter.getClass().getName()));
+
+    GenericRecord genericGmce = GenericData.get().deepCopy(gmceWithMockWriter.getSchema(), gmceWithMockWriter);
```
Comment (homatthew) [inline gobblin-iceberg/src/test/java/org/apache/gobblin/iceberg/writer/IcebergMetadataWriterTest.java:355]: This is the only substance change in the file. Add mock writer to the allow list in the GMCE

File: gobblin-iceberg/src/test/java/org/apache/gobblin/iceberg/writer/GobblinMCEWriterTest.java [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+    writeToMetadataWriters(gmceBuilder.build());
+
+    Mockito.verify(mockWriter, Mockito.times(1)).writeEnvelope(
+        Mockito.any(RecordEnvelope.class), Mockito.anyMap(), Mockito.anyMap(), Mockito.any(HiveSpec.class));
+    Mockito.verify(exceptionWriter, never()).writeEnvelope(
+        Mockito.any(RecordEnvelope.class), Mockito.anyMap(), Mockito.anyMap(), Mockito.any(HiveSpec.class));
+  }
+
+  @Test
+  public void testFaultTolerance() throws IOException {
```
Comment (homatthew) [inline gobblin-iceberg/src/test/java/org/apache/gobblin/iceberg/writer/GobblinMCEWriterTest.java:158]: This test kind of already exists in the Iceberg writer test but I felt the testing could be done more explicitly here without the fluff related to iceberg / hive

File: gobblin-iceberg/src/test/java/org/apache/gobblin/iceberg/writer/GobblinMCEWriterTest.java [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+    return new Object[][] {
+        {Arrays.asList(mockWriter.getClass().getName(), exceptionWriter.getClass().getName())},
+        {Collections.emptyList()}
+    };
+  }
+
+  private void initMocks() {
+    MockitoAnnotations.initMocks(this);
+    mockWriter = Mockito.mock(MetadataWriter.class);
+    exceptionWriter = Mockito.mock(HiveMetadataWriter.class);
```
Comment (homatthew) [inline gobblin-iceberg/src/test/java/org/apache/gobblin/iceberg/writer/GobblinMCEWriterTest.java:229]: This is kind of a hack so I could have 2 `MetadataWriter` mocks with different class names

**Review by ZihanLi58 — APPROVED** [review]

**Review by ZihanLi58 — APPROVED** [review]
LGTM, thanks for the change
